Fix domain matching in Route53DependencyLambda

komazarari opened the pull request:
Hi, I faced a problem while creating a stack.

my environment is, such like,
- CFn param DomainName: hubs.my.domain.com
- Route53 zone: my.domain.com

and I got the following error.
```
2020-04-16 17:39:22 UTC+0900	ExternalZoneInfo	
CREATE_FAILED	Failed to create resource. undefined See details in CloudWatch Log: 2020/04/16/[$LATEST]bf8dde3ec87246b8b899fdc0263ed32f

undefined See details in CloudWatch Log
Route53DependencyLambda
```
commit 6f51c92 fixes this.

The commit 5b6d9b5 is for another problem, I think, that `DomainName` is validated by pattern 
https://github.com/mozilla/hubs-ops/blob/ef92c35f876bc8afd963d199f799d45335c1a052/cloudformation/stack.yaml#L150 which allows a period at the end, but it may cause a matching failure.
```
> 'hubs.my.domain.com.'.split('.')
[ 'hubs', 'my', 'domain', 'com', '' ]
```

gfodor commented:
Thank you! 🙏 

I think this should be fixed in the originating project: https://github.com/sonyxperiadev/amazon-custom-resources/tree/master/route53-dependency

I'd prefer that the original author check this commit just for sanity's sake before merging. If they aren't able to, I will review + test directly. Thanks again for the contribution!

komazarari commented:
@gfodor Thank you so much for your quick comment!

But, sorry, I have a little bit of confusion. There are some differences between the lambda source in this repo and the original, and the source parts I wished to updates have no corresponding parts in the original. It seems to be added by https://github.com/mozilla/hubs-ops/commit/959f3d4f337293166de240e0426fb5c5d884f4fa . 
I'm sorry to trouble you, but could you review it?

komazarari commented:
Hi, @gfodor 

It's been a while, so I'll summarize this PR again.
Our original problem is encountering an error when creating a stack with the parameters:
- DomainName: hubs.my.domain.com
- Route53 zone: my.domain.com.
 
I think the `Route53DependencyLambda` causes that issue. 
In your previous comment, you pointed out an original project for the function. Still, I created this PR here because the part I'd like to update is not in the original repository but was added in this "hubs-ops" repository.

I've created simple test cases for this change that works in Nodejs without any test framework. It would be great if you cloud review it again. :)

```javascript
let properties;
let zone;

function machingFilterFunctionBefore(zone) {
  if (properties.Id) {
    return zone.Id === "/hostedzone/" + properties.Id;
  } else {
    var tldParts = properties.Domain.split(".");
    var tld = tldParts[tldParts.length - 2] + "." + tldParts[tldParts.length - 1];
    return zone.Name === tld + ".";
  }
}

function machingFilterFunctionAfter(zone) {
  if (properties.Id) {
    return zone.Id === "/hostedzone/" + properties.Id;
  } else {
    var domain = properties.Domain.slice(-1) === "."
        ? properties.Domain
        : properties.Domain + "."
    if (zone.Name === domain)
      return true;

    var tldParts = domain.split(".").filter(Boolean);
    var tld = tldParts.slice(1, tldParts.length).join(".");
    return zone.Name === tld + ".";
  }
}

//// please edit
//const testSubject = machingFilterFunctionBefore;
const testSubject = machingFilterFunctionAfter;
////

const testCases = [
  // https://hubs.mozilla.com/docs/hubs-cloud-aws-domain-recipes.html#recipe-1-dedicated-domain-on-route-53
  { r53zone: 'myhub.com.', hubsdomain: 'myhub.com', expected: true },
  { r53zone: 'myhub.com.', hubsdomain: 'myhub.com.', expected: true },
  { r53zone: 'myhub.com.', hubsdomain: 'hub.myhub.com', expected: true },
  { r53zone: 'myhub.com.', hubsdomain: 'hub.myhub.com.', expected: true },
  { r53zone: 'myhub.com.', hubsdomain: 'mysite.com', expected: false },
  { r53zone: 'myhub.com.', hubsdomain: 'mysite.com.', expected: false },
  { r53zone: 'myhub.com.', hubsdomain: 'hub.mysite.com', expected: false },
  { r53zone: 'myhub.com.', hubsdomain: 'hub.mysite.com.', expected: false },

  // https://hubs.mozilla.com/docs/hubs-cloud-aws-domain-recipes.html#recipe-2-domain-is-in-use-configure-subdomain-for-hub-on-route-53
  { r53zone: 'mysite.com.', hubsdomain: 'hub.mysite.com', expected: true },
  { r53zone: 'mysite.com.', hubsdomain: 'hub.mysite.com.', expected: true },
  { r53zone: 'mysite.com.', hubsdomain: 'hub.example.com', expected: false },
  { r53zone: 'mysite.com.', hubsdomain: 'hub.example.com.', expected: false },

  { r53zone: 'mydomain.mysite.com.', hubsdomain: 'hub.mydomain.mysite.com', expected: true },
  { r53zone: 'mydomain.mysite.com.', hubsdomain: 'hub.mydomain.mysite.com.', expected: true },
  { r53zone: 'mydomain.mysite.com.', hubsdomain: 'hub.mydomain.example.com', expected: false },
  { r53zone: 'mydomain.mysite.com.', hubsdomain: 'hub.mydomain.example.com.', expected: false },
  { r53zone: 'mydomain.mysite.com.', hubsdomain: 'hub.anotherdomain.mysite.com', expected: false },
  { r53zone: 'mydomain.mysite.com.', hubsdomain: 'hub.anotherdomain.mysite.com.', expected: false },

  { r53zone: 'some.mydomain.mysite.com.', hubsdomain: 'hub.some.mydomain.mysite.com', expected: true },
  { r53zone: 'some.mydomain.mysite.com.', hubsdomain: 'hub.some.mydomain.mysite.com.', expected: true },
  { r53zone: 'some.mydomain.mysite.com.', hubsdomain: 'hub.other.mydomain.mysite.com', expected: false },
  { r53zone: 'some.mydomain.mysite.com.', hubsdomain: 'hub.other.mydomain.mysite.com.', expected: false },
]

testCases.forEach((testCase, idx) => {
  /* setUp */
  properties = {
    Id: null,
    Domain: testCase.hubsdomain,
  };
  zone = { Name: testCase.r53zone };

  /* test */
  const result = testSubject(zone);
  if (result === testCase.expected) {
    console.log(`[${idx}] passed: 'Route53 zone: ${testCase.r53zone}' - 'DomainName: ${testCase.hubsdomain}' : expected to ${testCase.expected}`);
  } else {
    console.log(`[${idx}] failed: The pair 'Route53 zone: ${testCase.r53zone}' and 'DomainName: ${testCase.hubsdomain}' is expected to ${testCase.expected}.`);
  }
});
```

Before:
```
❯ node ~/test.js
[0] passed: 'Route53 zone: myhub.com.' - 'DomainName: myhub.com' : expected to true
[1] failed: The pair 'Route53 zone: myhub.com.' and 'DomainName: myhub.com.' is expected to true.
[2] passed: 'Route53 zone: myhub.com.' - 'DomainName: hub.myhub.com' : expected to true
[3] failed: The pair 'Route53 zone: myhub.com.' and 'DomainName: hub.myhub.com.' is expected to true.
[4] passed: 'Route53 zone: myhub.com.' - 'DomainName: mysite.com' : expected to false
[5] passed: 'Route53 zone: myhub.com.' - 'DomainName: mysite.com.' : expected to false
[6] passed: 'Route53 zone: myhub.com.' - 'DomainName: hub.mysite.com' : expected to false
[7] passed: 'Route53 zone: myhub.com.' - 'DomainName: hub.mysite.com.' : expected to false
[8] passed: 'Route53 zone: mysite.com.' - 'DomainName: hub.mysite.com' : expected to true
[9] failed: The pair 'Route53 zone: mysite.com.' and 'DomainName: hub.mysite.com.' is expected to true.
[10] passed: 'Route53 zone: mysite.com.' - 'DomainName: hub.example.com' : expected to false
[11] passed: 'Route53 zone: mysite.com.' - 'DomainName: hub.example.com.' : expected to false
[12] failed: The pair 'Route53 zone: mydomain.mysite.com.' and 'DomainName: hub.mydomain.mysite.com' is expected to true.
[13] failed: The pair 'Route53 zone: mydomain.mysite.com.' and 'DomainName: hub.mydomain.mysite.com.' is expected to true.
[14] passed: 'Route53 zone: mydomain.mysite.com.' - 'DomainName: hub.mydomain.example.com' : expected to false
[15] passed: 'Route53 zone: mydomain.mysite.com.' - 'DomainName: hub.mydomain.example.com.' : expected to false
[16] passed: 'Route53 zone: mydomain.mysite.com.' - 'DomainName: hub.anotherdomain.mysite.com' : expected to false
[17] passed: 'Route53 zone: mydomain.mysite.com.' - 'DomainName: hub.anotherdomain.mysite.com.' : expected to false
[18] failed: The pair 'Route53 zone: some.mydomain.mysite.com.' and 'DomainName: hub.some.mydomain.mysite.com' is expected to true.
[19] failed: The pair 'Route53 zone: some.mydomain.mysite.com.' and 'DomainName: hub.some.mydomain.mysite.com.' is expected to true.
[20] passed: 'Route53 zone: some.mydomain.mysite.com.' - 'DomainName: hub.other.mydomain.mysite.com' : expected to false
[21] passed: 'Route53 zone: some.mydomain.mysite.com.' - 'DomainName: hub.other.mydomain.mysite.com.' : expected to false
```

After:
```
❯ node ~/test.js
[0] passed: 'Route53 zone: myhub.com.' - 'DomainName: myhub.com' : expected to true
[1] passed: 'Route53 zone: myhub.com.' - 'DomainName: myhub.com.' : expected to true
[2] passed: 'Route53 zone: myhub.com.' - 'DomainName: hub.myhub.com' : expected to true
[3] passed: 'Route53 zone: myhub.com.' - 'DomainName: hub.myhub.com.' : expected to true
[4] passed: 'Route53 zone: myhub.com.' - 'DomainName: mysite.com' : expected to false
[5] passed: 'Route53 zone: myhub.com.' - 'DomainName: mysite.com.' : expected to false
[6] passed: 'Route53 zone: myhub.com.' - 'DomainName: hub.mysite.com' : expected to false
[7] passed: 'Route53 zone: myhub.com.' - 'DomainName: hub.mysite.com.' : expected to false
[8] passed: 'Route53 zone: mysite.com.' - 'DomainName: hub.mysite.com' : expected to true
[9] passed: 'Route53 zone: mysite.com.' - 'DomainName: hub.mysite.com.' : expected to true
[10] passed: 'Route53 zone: mysite.com.' - 'DomainName: hub.example.com' : expected to false
[11] passed: 'Route53 zone: mysite.com.' - 'DomainName: hub.example.com.' : expected to false
[12] passed: 'Route53 zone: mydomain.mysite.com.' - 'DomainName: hub.mydomain.mysite.com' : expected to true
[13] passed: 'Route53 zone: mydomain.mysite.com.' - 'DomainName: hub.mydomain.mysite.com.' : expected to true
[14] passed: 'Route53 zone: mydomain.mysite.com.' - 'DomainName: hub.mydomain.example.com' : expected to false
[15] passed: 'Route53 zone: mydomain.mysite.com.' - 'DomainName: hub.mydomain.example.com.' : expected to false
[16] passed: 'Route53 zone: mydomain.mysite.com.' - 'DomainName: hub.anotherdomain.mysite.com' : expected to false
[17] passed: 'Route53 zone: mydomain.mysite.com.' - 'DomainName: hub.anotherdomain.mysite.com.' : expected to false
[18] passed: 'Route53 zone: some.mydomain.mysite.com.' - 'DomainName: hub.some.mydomain.mysite.com' : expected to true
[19] passed: 'Route53 zone: some.mydomain.mysite.com.' - 'DomainName: hub.some.mydomain.mysite.com.' : expected to true
[20] passed: 'Route53 zone: some.mydomain.mysite.com.' - 'DomainName: hub.other.mydomain.mysite.com' : expected to false
[21] passed: 'Route53 zone: some.mydomain.mysite.com.' - 'DomainName: hub.other.mydomain.mysite.com.' : expected to false
```

Thank you!

komazarari commented:
I think this also fix the issue pointed on #103.
